src/v/kafka/server/quota_manager.cc

```diff
@@ -162,8 +231,11 @@ std::chrono::milliseconds quota_manager::do_record_partition_mutations(
       "This method can only be executed from quota manager home shard");
 
     auto quota_id = get_client_quota_id(client_id, {});
-    auto it = maybe_add_and_retrieve_quota(quota_id, now);
-    const auto units = it->second.pm_rate->record_and_measure(mutations, now);
+    auto cq = co_await maybe_add_and_retrieve_quota(quota_id, now);
+    if (!cq) {
+        co_return std::chrono::milliseconds::zero();
+    }
+    const auto units = cq.value()->pm_rate->record_and_measure(mutations, now);
```

Comment:
As discussed, we can probably void the copy of what was previously a seastar shared_pointer by moving to an interface that encapsulates this:
```c++
    return record_and_measure_pm(quota_id, mutations, now);
```

Reply:
I've pushed more changes now for this as well: https://github.com/redpanda-data/redpanda/pull/18098#issuecomment-2124940958.

The interface for this is really the interface of the `quota_manager`, eg. `record_partition_mutations(client_id, mutations, now)`. I haven't found a better interface to wrap the map in but I am open to suggestions.